enhancements/machine-api/control-plane.md

```diff
@@ -0,0 +1,281 @@
+
+	// ErrorMessage indicates that there is a terminal problem reconciling the
+	// state, and will be set to a descriptive error message.
+	// +optional
+	FailureMessage *string `json:"failureMessage,omitempty"`
+}
+
+type ControlPlaneStatusError string
+```
+
+### Risks and Mitigations
+
+During horizontal scaling operations there are sensitive scenarios like scaling from 1 to 2. As soon as the Etcd API is notified of the new member the cluster loses quorum until that new member starts and joins the cluster. This must be still handled by the [Cluster Etcd Operator](https://github.com/openshift/enhancements/blob/master/enhancements/etcd/cluster-etcd-operator.md#motivation) while the Control Plane controller should honour and short-circuit when it meets the Etcd unhealthiness criteria as described in the workflows above.
+
+There are multiple components and operators involved in the lifecycle management of the Control Plane. This in addition to its complex nature makes it difficult to predict the holistic behaviour for failure scenarios. To mitigate risks the Control Plane controller will short-circuit preventing any scaling operation until its healthiness criteria is met again.
+
+## Design Details
+
+### Test Plan
+
+- Exhaustive unit testing.
+- Exhaustive integration testing via [envTest](https://book.kubebuilder.io/reference/testing/envtest.html).
+- E2e testing on the machine API [e2e test suite](https://github.com/openshift/cluster-api-actuator-pkg/tree/master/pkg):
+	- Scale from 3 to 5.
+	- Scale from 5 to 3.
+	- Vertical scaling Rolling upgrade.
+	- Simulate 1 unhealthy node and remediation.
+- E2e testing on the origin e2e test suite.
+
+### Graduation Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+This controller will go GA in 4.N as long as:
+- All the testing above is in place.
+- There is considerable internal usage and manual disruptive QE testing feedback.
+- This is beta tested pre-release during a considerable period of time either internally or by an early adopter user through the dev preview builds.
+
+Otherwise this would go "Tech Preview" under a feature gate first until enough real use feedback is developed. The risk though is that there might be a very scarce use of feature gated features in the field as to develop the feedback we need.
+
+### Upgrade / Downgrade Strategy
+
+During the cluster upgrade for the targeted release the Machine API Operator (MAO) will let the CVO to instantiate the new CRD `controlPlane` and it will run the backing controller making this functionallity opt-in for existing clusters. The user can create an instance of the new CRD if they choose to do so.
+
+New IPI clusters deployed after the targeted release will run the `controlPlane` instance deployed by the installer out of the box.
+
+### Version Skew Strategy
+
+## Implementation History
+- "https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20191017-kubeadm-based-control-plane.md"
```

Comment:
How similar is this to upstream? Given we want to eventually converge somewhat with upstream, does this fall in line with the assumptions for a Control Plane upstream?

Reply:
This proposal is fairly aligned with the capi kubeadm control plane one. The main difference is here we don't need to manage the lifecycle of etcd via kubeadm. We just want to check its healthy as a gate for proceeding with scale operations.
See also https://github.com/kubernetes-sigs/cluster-api/issues/2836